src/IndexMaps/abstractindexmap.jl

```diff
@@ -6,19 +6,17 @@ using ITensorNetworks: IndsNetwork, vertex_data
 abstract type AbstractIndexMap{VB,VD} end
 
 #These functions need to be defined on the concrete type for implementation
-index_digit(imap::AbstractIndexMap) = not_implemented()
-index_dimension(imap::AbstractIndexMap) = not_implemented()
-Base.copy(imap::AbstractIndexMap) = not_implemented()
-index_value_to_scalar(imap::AbstractIndexMap, ind::Index, value::Int) = not_implemented()
-ITensors.inds(imap::AbstractIndexMap) = not_implemented()
-ind(imap::AbstractIndexMap, args...) = not_implemented()
+function index_digit(imap::AbstractIndexMap) end
```

Comment:
This creates a function that returns `nothing`:
```julia
julia> function f(x::Int) end
f (generic function with 1 method)

julia> f(3)

julia> isnothing(f(3))
true
```
As an alternative, you could manually `throw` a `MethodError`: https://docs.julialang.org/en/v1/base/base/#Core.MethodError.

Reply:
I see the subtlety now, thanks, I'll rewrite a few of these

Reply:
After consultation with Karl, it seems the right move is to do
`function index_digit end`
And then not define `ITensors.inds` etc. This is a bit weird to me but I think its the most Julia way, although we can always define an error

Reply:
Hmm `function index_digit end` still defines the function `index_digit` though which now has no methods but can be called which I find a bit weird.

The suggestion in the post you linked at the start seems reasonable?. I.e. why don't we just remove the `not_implemented` and just add the list of functions which need to be defined on the concrete type as a comment?

Reply:
The blog post recommends this "empty" function, so that you get a type callback when something isn't implemented. It's used in ITensors too, see here for example: https://github.com/ITensor/ITensors.jl/blob/main/NDTensors/src/lib/TypeParameterAccessors/src/interface.jl

I agree its weird but I think this lets a doc system pick up the interface rather than just a big comment?

Reply:
Interesting okay that makes sense. So we can do that for all of them other than ones where we overload functions?